src/mockFA2.ml

```diff
@@ -117,21 +118,26 @@ let[@inline] fa2_run_update_operators
     xs
 
 let[@inline] update_operators (state: mock_fa2_state) (xs: fa2_update_operator list) : LigoOp.operation list * mock_fa2_state =
-  let { fa2_state = fa2_state; metadata = metadata; } = state in (* deconstruct *)
+  let { fa2_state = fa2_state; total_token = total_token; metadata = metadata; } = state in (* deconstruct *)
   let fa2_state = fa2_run_update_operators (fa2_state, xs) in
-  let state = { fa2_state = fa2_state; metadata = metadata; } in (* reconstruct *)
+  let state = { fa2_state = fa2_state; total_token = total_token; metadata = metadata; } in (* reconstruct *)
   (([]: LigoOp.operation list), state)
 
 let[@inline] mint (state: mock_fa2_state) (amnt: Ligo.nat) : LigoOp.operation list * mock_fa2_state =
-  let { fa2_state = fa2_state; metadata = metadata; } = state in (* deconstruct *)
+  let { fa2_state = fa2_state; total_token = total_token; metadata = metadata; } = state in (* deconstruct *)
   let fa2_state = ledger_issue (fa2_state, mock_fa2_token_id, !Ligo.Tezos.sender, amnt) in
-  let state = { fa2_state = fa2_state; metadata = metadata; } in (* reconstruct *)
+  let total_token = Ligo.add_nat_nat total_token amnt in
+  let state = { fa2_state = fa2_state; total_token = total_token; metadata = metadata; } in (* reconstruct *)
   (([]: LigoOp.operation list), state)
 
 let[@inline] redeem (state: mock_fa2_state) (amnt: Ligo.nat) : LigoOp.operation list * mock_fa2_state =
-  let { fa2_state = fa2_state; metadata = metadata; } = state in (* deconstruct *)
+  let { fa2_state = fa2_state; total_token = total_token; metadata = metadata; } = state in (* deconstruct *)
   let fa2_state = ledger_withdraw (fa2_state, mock_fa2_token_id, !Ligo.Tezos.sender, amnt) in
-  let state = { fa2_state = fa2_state; metadata = metadata; } in (* reconstruct *)
+  let total_token =
+    match Ligo.is_nat (Ligo.sub_nat_nat total_token amnt) with
+    | None -> (failwith "FA2_INSUFFICIENT_BALANCE" : Ligo.nat)
```

Comment:
I'm wondering if it's okay to throw this error here or if we need something more specific. I couldn't find any relevant details in TZIP-12 other than that this should be thrown when transfers exceed the senders balance in the ledger.

Reply:
Right. So, since this is in entrypoint code (not in metadata), TZIP-12 is oblivious to the specific error we throw (and, indeed, TZIP-16 does not specify any error we MUST emit). TZIP-12 does though. Currently I i
mplemented all additions/subtractions in this order:
```
account[user] -= amnt
total_token   -= amnt
```
and not
```
total_token   -= amnt
account[user] -= amnt
```
But I didn't want the code to be sensitive to the order of the two subtractions, so I made them both fail with the correct error mnemonic (i.e., so that we comply with TZIP12 independently of the order of these subtractions). Alternative we could make the `total_token` reduction fail with an internal "impossible-happened" error, but then we'd have to always perform that subtraction **after** the subtraction from the specific account.

Reply:
Gotcha, thanks for the additional background. In that case I'm also happy leaving it as-is. I don't think there is really a huge benefit to throwing a different error code here other than to indicate that the error is an internal one.